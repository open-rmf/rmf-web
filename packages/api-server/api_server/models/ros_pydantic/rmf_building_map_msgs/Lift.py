# This is a generated file, do not edit

from typing import List

import pydantic

from ..rmf_building_map_msgs.Door import Door
from ..rmf_building_map_msgs.Graph import Graph


class Lift(pydantic.BaseModel):
    name: str = ""  # string
    levels: List[str] = []  # string
    doors: List[Door] = []  # rmf_building_map_msgs/Door
    wall_graph: Graph = Graph()  # rmf_building_map_msgs/Graph
    ref_x: float = 0  # float32
    ref_y: float = 0  # float32
    ref_yaw: float = 0  # float32
    width: float = 0  # float32
    depth: float = 0  # float32

    class Config:
        orm_mode = True
        schema_extra = {
            "required": [
                "name",
                "levels",
                "doors",
                "wall_graph",
                "ref_x",
                "ref_y",
                "ref_yaw",
                "width",
                "depth",
            ],
        }


# string name
# string[] levels
# Door[] doors
# Graph wall_graph
#
# # (ref_x, ref_y, ref_yaw) is a "reference orientation" of the lift cabin
# # which can be used to align floors.
# float32 ref_x
# float32 ref_y
# float32 ref_yaw
#
# # width and depth of the cabin
# float32 width
# float32 depth
