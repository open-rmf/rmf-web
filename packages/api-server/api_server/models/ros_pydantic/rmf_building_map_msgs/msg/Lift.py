# This is a generated file, do not edit

from typing import Annotated

import pydantic

from .Door import Door as rmf_building_map_msgs_msg_Door
from .Graph import Graph as rmf_building_map_msgs_msg_Graph


class Lift(pydantic.BaseModel):
    model_config = pydantic.ConfigDict(from_attributes=True)

    name: str
    levels: list[str]
    doors: list[rmf_building_map_msgs_msg_Door]
    wall_graph: rmf_building_map_msgs_msg_Graph
    ref_x: float
    ref_y: float
    ref_yaw: float
    width: float
    depth: float
