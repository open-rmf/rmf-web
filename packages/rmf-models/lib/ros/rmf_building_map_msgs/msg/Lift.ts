/* This is a generated file, do not edit */

import { Door } from '../../rmf_building_map_msgs/msg/Door';
import { Graph } from '../../rmf_building_map_msgs/msg/Graph';

export class Lift {
  static readonly FullTypeName = 'rmf_building_map_msgs/msg/Lift';

  name: string;
  levels: string[];
  doors: Door[];
  wall_graph: Graph;
  ref_x: number;
  ref_y: number;
  ref_yaw: number;
  width: number;
  depth: number;

  constructor(fields: Partial<Lift> = {}) {
    this.name = fields.name || '';
    this.levels = fields.levels || [];
    this.doors = fields.doors || [];
    this.wall_graph = fields.wall_graph || new Graph();
    this.ref_x = fields.ref_x || 0;
    this.ref_y = fields.ref_y || 0;
    this.ref_yaw = fields.ref_yaw || 0;
    this.width = fields.width || 0;
    this.depth = fields.depth || 0;
  }

  static validate(obj: Record<string, unknown>): void {
    if (typeof obj['name'] !== 'string') {
      throw new Error('expected "name" to be "string"');
    }
    if (!Array.isArray(obj['levels'])) {
      throw new Error('expected "levels" to be an array');
    }
    for (const [i, v] of obj['levels'].entries()) {
      if (typeof v !== 'string') {
        throw new Error(`expected index ${i} of "levels" to be "string"`);
      }
    }
    if (!Array.isArray(obj['doors'])) {
      throw new Error('expected "doors" to be an array');
    }
    for (const [i, v] of obj['doors'].entries()) {
      try {
        Door.validate(v);
      } catch (e) {
        throw new Error(`in index ${i} of "doors":\n  ` + (e as Error).message);
      }
    }
    try {
      Graph.validate(obj['wall_graph'] as Record<string, unknown>);
    } catch (e) {
      throw new Error('in "wall_graph":\n  ' + (e as Error).message);
    }
    if (typeof obj['ref_x'] !== 'number') {
      throw new Error('expected "ref_x" to be "number"');
    }
    if (typeof obj['ref_y'] !== 'number') {
      throw new Error('expected "ref_y" to be "number"');
    }
    if (typeof obj['ref_yaw'] !== 'number') {
      throw new Error('expected "ref_yaw" to be "number"');
    }
    if (typeof obj['width'] !== 'number') {
      throw new Error('expected "width" to be "number"');
    }
    if (typeof obj['depth'] !== 'number') {
      throw new Error('expected "depth" to be "number"');
    }
  }
}

/*
string name
string[] levels
Door[] doors
Graph wall_graph

# (ref_x, ref_y, ref_yaw) is a "reference orientation" of the lift cabin
# which can be used to align floors.
float32 ref_x
float32 ref_y
float32 ref_yaw

# width and depth of the cabin
float32 width
float32 depth
*/
