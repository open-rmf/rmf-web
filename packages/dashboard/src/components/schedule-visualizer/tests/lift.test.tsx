import { createMount } from '@material-ui/core/test-utils';
import * as RomiCore from '@osrf/romi-js-core-interfaces';
import fakeLiftStates from '../../../mock/data/lift-states';
import getBuildingMap from '../../../mock/data/building-map';
import Lift from '../lift';
import React from 'react';
import toJson from 'enzyme-to-json';

const mount = createMount();

const getLiftComponent = (
  lift: RomiCore.Lift,
  state: RomiCore.LiftState,
  currentFloor: string = 'L1',
): React.ReactElement => {
  return (
    <svg>
      <Lift currentFloor={currentFloor} lift={lift} liftState={state} />
    </svg>
  );
};

const buildWrapper = (
  lift: RomiCore.Lift,
  state: RomiCore.LiftState,
  currentFloor: string = 'L1',
) => {
  const wrapper = mount(getLiftComponent(lift, state, currentFloor));

  const liftSVGRect = wrapper
    .find(Lift)
    .at(0)
    .find('rect');

  return { wrapper, liftSVGRect };
};

const loadLift = async () => {
  const buildingMap = await getBuildingMap();
  const lifts = buildingMap.lifts;
  const lift = lifts[0];
  const state = fakeLiftStates()[lift.name];
  return { lift, state };
};

it('Trigger click event', async () => {
  const buildingMap = await getBuildingMap();
  const lifts = buildingMap.lifts;
  const lift = lifts[0];
  let clicked = false;

  const wrapper = mount(
    <svg>
      <Lift currentFloor={'L1'} lift={lift} onClick={() => (clicked = true)} />
    </svg>,
  );

  wrapper
    .find(Lift)
    .at(0)
    .simulate('click');
  expect(clicked).toBe(true);

  wrapper.unmount();
});

describe('Checks assignation of styles on different mode of the Lift', () => {
  test('Check orange color and _Fire_ text when the lift its on Fire mode', async () => {
    const { lift, state } = await loadLift();
    state.current_mode = RomiCore.LiftState.MODE_FIRE;
    const { wrapper, liftSVGRect } = buildWrapper(lift, state);
    expect(liftSVGRect.hasClass(/(makeStyles)-(fire)-(\d+)/)).toBe(true);
    expect(wrapper.find('#liftMode').text()).toEqual('FIRE!');
    wrapper.unmount();
  });

  test('Lift on Emergency mode', async () => {
    const { lift, state } = await loadLift();
    state.current_mode = RomiCore.LiftState.MODE_EMERGENCY;
    const { wrapper, liftSVGRect } = buildWrapper(lift, state);
    expect(liftSVGRect.hasClass(/(makeStyles)-(emergency)-(\d+)/)).toBe(true);
    expect(wrapper.find('#liftMode').text()).toEqual('EMERGENCY!');
    wrapper.unmount();
  });

  test('Lift on Disabled mode', async () => {
    const { lift, state } = await loadLift();
    state.current_mode = RomiCore.LiftState.MODE_OFFLINE;
    const { wrapper, liftSVGRect } = buildWrapper(lift, state);
    expect(liftSVGRect.hasClass(/(makeStyles)-(offLine)-(\d+)/)).toBe(true);
    expect(wrapper.find('#liftMode').text()).toEqual('OFFLINE');
    wrapper.unmount();
  });

  test('Lift on Unknown mode', async () => {
    const { lift, state } = await loadLift();
    state.current_mode = RomiCore.LiftState.MODE_UNKNOWN;
    const { wrapper, liftSVGRect } = buildWrapper(lift, state);
    expect(liftSVGRect.hasClass(/(makeStyles)-(unknownLift)-(\d+)/)).toBe(true);
    expect(wrapper.find('#liftMode').exists()).toBe(false);
    wrapper.unmount();
  });
});

describe('Checks assignation of styles on combination of motion states and mode of the Lift', () => {
  test('Lift its on current floor and its on Human mode', async () => {
    const { lift, state } = await loadLift();
    state.current_mode = RomiCore.LiftState.MODE_HUMAN;
    state.current_floor = 'L1';
    const { wrapper, liftSVGRect } = buildWrapper(lift, state);
    expect(liftSVGRect.hasClass(/(makeStyles)-(humanMode)-(\d+)/)).toBe(true);
    expect(wrapper.find('#liftMotion').text()).toEqual('L1');
    expect(wrapper.find('#liftMode').exists()).toBe(false);
    wrapper.unmount();
  });

  test('Lift its on current floor and its on SVG mode', async () => {
    const { lift, state } = await loadLift();
    state.current_mode = RomiCore.LiftState.MODE_AGV;
    state.current_floor = 'L1';
    const { wrapper, liftSVGRect } = buildWrapper(lift, state);
    expect(liftSVGRect.hasClass(/(makeStyles)-(liftOnCurrentFloor)-(\d+)/)).toBe(true);
    expect(wrapper.find('#liftMotion').text()).toEqual('L1');
    expect(wrapper.find('#liftMode').exists()).toBe(false);
    wrapper.unmount();
  });

  test('Lift its not on current floor and it`s going to an upper floor', async () => {
    const { lift, state } = await loadLift();
    state.current_floor = 'L3';
    state.destination_floor = 'L4';
    state.motion_state = RomiCore.LiftState.MOTION_UP;
    const { wrapper, liftSVGRect } = buildWrapper(lift, state);
    expect(liftSVGRect.hasClass(/(makeStyles)-(liftMoving)-(\d+)/)).toBe(true);
    expect(wrapper.find('#liftMotion').text()).toEqual('L3 → L4');
    expect(wrapper.find('#liftMode').exists()).toBe(false);
    wrapper.unmount();
  });

  test('Lift its not on current floor and it`s going to an lower floor', async () => {
    const { lift, state } = await loadLift();
    state.current_floor = 'L4';
    state.destination_floor = 'L2';
    state.motion_state = RomiCore.LiftState.MOTION_DOWN;
    const { wrapper, liftSVGRect } = buildWrapper(lift, state);
    expect(liftSVGRect.hasClass(/(makeStyles)-(liftMoving)-(\d+)/)).toBe(true);
    expect(wrapper.find('#liftMotion').text()).toEqual(`L4 → L2`);
    expect(wrapper.find('#liftMode').exists()).toBe(false);
    wrapper.unmount();
  });

  test('Lift its not on current floor and its mode is UNKNOWN', async () => {
    const { lift, state } = await loadLift();
    state.current_floor = 'L4';
    state.destination_floor = 'L2';
    state.current_mode = RomiCore.LiftState.MODE_HUMAN;
    state.motion_state = RomiCore.LiftState.MOTION_DOWN;
    const { wrapper, liftSVGRect } = buildWrapper(lift, state);
    expect(liftSVGRect.hasClass(/(makeStyles)-(liftMoving)-(\d+)/)).toBe(true);
    expect(wrapper.find('#liftMotion').text()).toEqual(`L4 → L2`);
    expect(wrapper.find('#liftMode').exists()).toBe(false);
    wrapper.unmount();
  });

  test('Lift its on current floor and its mode is UNKNOWN', async () => {
    const { lift, state } = await loadLift();
    state.current_floor = 'L1';
    state.current_mode = RomiCore.LiftState.MODE_UNKNOWN;
    state.motion_state = RomiCore.LiftState.MOTION_STOPPED;
    const { wrapper, liftSVGRect } = buildWrapper(lift, state);
    expect(liftSVGRect.hasClass(/(makeStyles)-(unknownLift)-(\d+)/)).toBe(true);
    expect(wrapper.find('#liftMotion').text()).toEqual(`L1`);
    expect(wrapper.find('#liftMode').exists()).toBe(false);
    wrapper.unmount();
  });
});

test('Lift renders correctly', async () => {
  const { lift, state } = await loadLift();
  state.current_floor = 'L1';
  state.current_mode = RomiCore.LiftState.MODE_FIRE;
  state.motion_state = RomiCore.LiftState.MOTION_STOPPED;
  const { wrapper } = buildWrapper(lift, state);
  expect(toJson(wrapper)).toMatchSnapshot();
});
