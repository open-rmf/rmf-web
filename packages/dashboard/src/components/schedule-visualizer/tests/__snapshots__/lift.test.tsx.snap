// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lift renders correctly 1`] = `
<svg>
  <ForwardRef
    currentFloor="L1"
    lift={
      Object {
        "depth": 2.5,
        "doors": Array [
          Object {
            "door_type": 1,
            "motion_direction": 1,
            "motion_range": 0,
            "name": "lift1_front_door",
            "v1_x": 8.8,
            "v1_y": -2.3,
            "v2_x": 7.7,
            "v2_y": -4.5,
          },
        ],
        "levels": Array [
          "L1",
          "L2",
          "L3",
        ],
        "name": "Lift1",
        "ref_x": 7.1,
        "ref_y": -2.8,
        "ref_yaw": 0.5,
        "wall_graph": Object {
          "edges": Array [],
          "name": "wallgraph",
          "params": Array [],
          "vertices": Array [],
        },
        "width": 2.5,
      }
    }
    liftState={
      Object {
        "available_floors": Array [
          "L1",
          "L2",
          "L3",
        ],
        "available_modes": Uint8Array [
          2,
        ],
        "current_floor": "L1",
        "current_mode": 3,
        "destination_floor": "L1",
        "door_state": 2,
        "lift_name": "Lift1",
        "lift_time": Object {
          "nanosec": 0,
          "sec": 0,
        },
        "motion_state": 0,
        "session_id": "",
      }
    }
  >
    <g
      onClick={[Function]}
    >
      <rect
        className="makeStyles-liftMarker-191 makeStyles-lift-192 makeStyles-fire-197"
        height={2.5}
        rx="0.1"
        ry="0.1"
        transform="rotate(28.64788975654116, 7.1,2.8)"
        width={2.5}
        x={5.85}
        y={1.5499999999999998}
      />
      <text
        className="makeStyles-liftText-200"
        id="liftMotion"
        x={7.1}
        y={2.175}
      >
        L1
      </text>
      <text
        className="makeStyles-liftText-200"
        id="liftMode"
        x={7.1}
        y={2.8}
      >
        FIRE!
      </text>
    </g>
    <ForwardRef
      currentMode={2}
      door={
        Object {
          "door_type": 1,
          "motion_direction": 1,
          "motion_range": 0,
          "name": "lift1_front_door",
          "v1_x": 8.8,
          "v1_y": -2.3,
          "v2_x": 7.7,
          "v2_y": -4.5,
        }
      }
      key="lift-door-lift1_front_door"
    >
      <g
        aria-label="lift1_front_door"
        data-component="Door"
      >
        <SingleSlideDoor
          currentMode={2}
          door={
            Object {
              "door_type": 1,
              "motion_direction": 1,
              "motion_range": 0,
              "name": "lift1_front_door",
              "v1_x": 8.8,
              "v1_y": -2.3,
              "v2_x": 7.7,
              "v2_y": -4.5,
            }
          }
          onClick={[Function]}
          v1={
            Array [
              8.8,
              -2.3,
            ]
          }
          v2={
            Array [
              7.7,
              -4.5,
            ]
          }
        >
          <DefaultDoor
            style="makeStyles-doorOpen-204"
            v1={
              Array [
                8.8,
                2.3,
              ]
            }
            v2={
              Array [
                7.7,
                4.5,
              ]
            }
          >
            <g
              onClick={[Function]}
            >
              <line
                className="makeStyles-doorMarker-201 makeStyles-door-202 makeStyles-doorOpen-204"
                x1={8.8}
                x2={7.7}
                y1={2.3}
                y2={4.5}
              />
            </g>
          </DefaultDoor>
          <DefaultDoor
            onClick={[Function]}
            style="makeStyles-doorTransparent-207"
            v1={
              Array [
                8.8,
                2.3,
              ]
            }
            v2={
              Array [
                7.7,
                4.5,
              ]
            }
          >
            <g
              onClick={[Function]}
            >
              <line
                className="makeStyles-doorMarker-201 makeStyles-door-202 makeStyles-doorTransparent-207"
                x1={8.8}
                x2={7.7}
                y1={2.3}
                y2={4.5}
              />
            </g>
          </DefaultDoor>
        </SingleSlideDoor>
      </g>
    </ForwardRef>
  </ForwardRef>
</svg>
`;
